### PR TITLE
Box3: new method intersectsTriangle

### PR DESCRIPTION
**Pull request by matthias-w**
The PR adds the method `Box3.intersectsTriangle` #13064. The method tests whether the axis aligned box intersects with the given triangle. The algorithm used is based on separating axis theorem (SAT). This test is useful in computational geometry. I use it for generating a spatial index for `THREE.BufferGeometry`.

### CODE REVIEW

**Comment by mrdoob** [comment]
Would you like to also add a example that uses this?

**Comment by Mugen87** [comment]
We also need these things:

- Updated [docs](https://threejs.org/docs/index.html#api/math/Box3)
- Updated [unit test](https://github.com/mrdoob/three.js/blob/dev/test/unit/src/math/Box3.tests.js)
- Better code style. The code is not entirely conform to the project's [code style](https://github.com/mrdoob/three.js/wiki/Mr.doob%27s-Code-Style%E2%84%A2).
- We also should add the counter method `Triangle.intersectsBox()` (and update the corresponding doc page and unit test)

File: src/math/Box3.js [hunk 1]
```diff
@@ -371,6 +371,107 @@ Object.assign( Box3.prototype, {
...
+		var f2 = new Vector3();
+
+		var testAxis = new Vector3();
+
+		var center = new Vector3();
+		var extents = new Vector3();
+
+		var triangleNormal = new Vector3();
+
+		function satForAxes (axes) {
```
Comment (Mugen87) [inline src/math/Box3.js]: We should move this function to module space, see 166ba2f5e6d69da5efb8a9c214f59a1a18ce0ee3

**Comment by matthias-w** [comment]
@Mugen87 I will see to your mentioned open issues. Just one question: Can you please give me a hint regarding code style issues? I used the online code beautifier so I thought the code would be compliant to the required style rules.

@mrdoob Do you mean an example for a specific use case for the [THREE example page](https://threejs.org/examples/)? The only use case I currently have is the mentioned spatial index creation (Octree) for `THREE.BufferGeometry`. I use this approach for speeding up VRController-Mesh collision detection in my VR application.

File: src/math/Box3.js (hunk 1 above)
Comment (Mugen87) [inline src/math/Box3.js]:
```javascript
function satForAxes( axes ) {
```

File: src/math/Box3.js [hunk 2]
```diff
@@ -371,6 +371,107 @@ Object.assign( Box3.prototype, {
...
+
+				}
+
+			}
+
+			return true;
+
+		}
+
+		return function intersectsTriangle ( triangle ) {
```
Comment (Mugen87) [inline src/math/Box3.js]:
```javascript
return function intersectsTriangle( triangle ) {
```

File: src/math/Box3.js [hunk 3]
```diff
@@ -371,6 +371,107 @@ Object.assign( Box3.prototype, {
...
+
+			if ( this.isEmpty() ) {
+
+				return false;
+
+			}
+
+			// compute box center and extents
+			this.getCenter(center);
```
Comment (Mugen87) [inline src/math/Box3.js]:
```javascript
this.getCenter( center );
```

**Review by Mugen87 — CHANGES_REQUESTED** [review]

**Comment by Mugen87** [comment]
👍

**Review by Mugen87 — APPROVED** [review]

**Comment by mrdoob** [comment]
> @mrdoob Do you mean an example for a specific use case for the [THREE example page](https://threejs.org/examples/)? The only use case I currently have is the mentioned spatial index creation (Octree) for `THREE.BufferGeometry`. I use this approach for speeding up VRController-Mesh collision detection in my VR application.  

Yeah. I think it'll be good to have some example using it so other people have a hint of what to do with it.

Thanks!

**Pull request merged**